File: crates/chidori-js/benchmarks/workloads/typed_array.js
// Typed-array element traffic — the numeric-buffer class (DSP, embeddings,
// image/audio scratch): fill, sum, dot product, and in-place transform over
// Float64Array plus an Int32Array bit-mix pass. Mirrors array_sum.js but on
// typed arrays, which the typed loop kernels do NOT yet accept as a base
// (element access only translates for dense `Internal::Array`), so this runs
// on the generic interpreter path today. Deterministic fill (no RNG) so every
// runtime computes the same checksum.
const N = 50_000;
const ROUNDS = 4;
const a = new Float64Array(N);
const b = new Float64Array(N);
for (let i = 0; i < N; i++) {
  a[i] = (i * 7919) % 10007;
  b[i] = (i * 104729) % 7919;
}
let checksum = 0;
for (let r = 0; r < ROUNDS; r++) {
  let s = 0;
  for (let i = 0; i < a.length; i++) {
    s += a[i];
  }
  let d = 0;
  for (let i = 0; i < a.length; i++) {
    d += a[i] * b[i];
  }
  for (let i = 0; i < a.length; i++) {
    a[i] = (a[i] + b[i]) % 10007;
  }
  checksum = (checksum + s + d) % 9007199254740991;
}
// Int32Array pass: integer wrap + shift semantics.
const m = new Int32Array(1024);
for (let i = 0; i < m.length; i++) {
  m[i] = (i * 2654435761) | 0;
}
let mix = 0;
for (let r = 0; r < 50; r++) {
  for (let i = 0; i < m.length; i++) {
    mix = (mix ^ m[i]) + ((mix << 5) | 0) | 0;
  }
}
console.log("RESULT=" + (checksum + (mix >>> 0)));

File: crates/chidori-js/benchmarks/workloads/typed_array.py
# Python twin of typed_array.js — numeric-buffer traffic over array('d')
# (CPython's unboxed float64 buffer, the stdlib analog of Float64Array):
# fill, sum, dot product, in-place transform, plus an int32 bit-mix pass.
#
# JS-semantics notes: the float64 work is IEEE doubles on both sides, so the
# checksum matches by construction (Python float % and JS % agree for
# positive operands). Int32Array wrap-around is reproduced with an explicit
# to_int32 — array('i') would raise OverflowError instead of wrapping, so the
# mix pass runs on plain ints kept in int32 range.
from array import array

N = 50_000
ROUNDS = 4
a = array("d", bytes(8 * N))
b = array("d", bytes(8 * N))
for i in range(N):
    a[i] = (i * 7919) % 10007
    b[i] = (i * 104729) % 7919
checksum = 0.0
for r in range(ROUNDS):
    s = 0.0
    for i in range(len(a)):
        s += a[i]
    d = 0.0
    for i in range(len(a)):
        d += a[i] * b[i]
    for i in range(len(a)):
        a[i] = (a[i] + b[i]) % 10007
    checksum = (checksum + s + d) % 9007199254740991


# Int32 pass: integer wrap + shift semantics (JS `| 0`).
def to_int32(x):
    return ((x & 0xFFFFFFFF) ^ 0x80000000) - 0x80000000


m = [0] * 1024
for i in range(len(m)):
    m[i] = to_int32(i * 2654435761)
mix = 0
for r in range(50):
    for i in range(len(m)):
        mix = to_int32((mix ^ m[i]) + to_int32(mix << 5))
print("RESULT=" + str(int(checksum + (mix % 4294967296))))
